### DISCUSSION TITLE
TypeScript integration?

### DISCUSSION THREAD

**Discussion by OnkelTem** [colinhacks/zod#1837] [Q&A] [answered]
I can't find this section in the docs.

For example, the corresponding section of the Yup docs:
https://github.com/jquense/yup#ensuring-a-schema-matches-an-existing-type

```ts
import { object, number, string, ObjectSchema } from 'yup';

interface Person {
  name: string;
  age?: number;
  sex: 'male' | 'female' | 'other' | null;
}

const schema: ObjectSchema<Person> = object({
  name: string().defined(),
  age: number().optional(),
  sex: string<'male' | 'female' | 'other'>().nullable().defined();
});
```

How to do the same with Zod?

**Comment by 1valdis** [comment]
```ts
import { z } from 'zod';

const schema = z.object({
  name: z.string(),
  age: z.number().optional(),
  sex: z.enum(['male', 'female', 'other']).nullable(),
});

type Person = z.infer<typeof schema>;
```

**Reply by OnkelTem** [reply]
So you infer the type from a schema. 
But if you look at my question you will see that it's not what I was asking about.
"TypeScript integration" implies (perhaps, not so unambiguously...) that you have some typescript and you need to integrate schema with it. I.e. you **already** have a type and you need to validate data against it. Not the reverse. 
I even quoted a snippet that does precisely that.

**Reply by 1valdis** [reply]
"TypeScript integration" only means that something works in TypeScript and is strongly typed. The original question contained a type and a schema in one file and my answer does exactly that too, a type and a schema. So I answered what was originally asked.

**Reply by OnkelTem** [reply]
No, you didn't.

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
interface Person {
    name: string
    age?: number
    sex: 'male' | 'female' | 'other' | null
}

const personSchema = z.object( {
    name: z.string(),
    age: z.number().optional(),
    sex: z.enum( [ 'male', 'female', 'other' ] ).nullable(),
} ) satisfies z.ZodType<Person>
```

**Reply by OnkelTem** [reply]
Yes. I think I found it already. Yesterday **Svish** @ Discord suggested this:

```ts
interface Test {
  name: string;
  age: number;
}

const Schema: z.ZodType<Test> = z.object({
  name: z.string(),
  age: z.number(),
})

const test = Schema.parse(null); // <-- gets type Test
```

which is precisely what I was looking for. However, I haven't yet tested it myself. Gonna do it later today.

**Reply by OnkelTem** [reply]
Wait a second. `satisfies`? I haven't seen anything like this before :)

Found docs here: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html

P.S. I wonder is it the same as: 

```ts
type Expect<T extends true> = T
type isMatches<T, P> = T extends P ? true : false

type _ = Expect<isMatches<z.infer<typeof personSchema>, Person>>
```

I.e.: 

```ts
interface Person {
    name: string
    age?: number
    sex: 'male' | 'female' | 'other' | null
}

const personSchema = z.object( {
    name: z.string(),
    age: z.number().optional(),
    sex: z.enum( [ 'male', 'female', 'other' ] ).nullable(),
} )

type _ = Expect<isMatches<z.infer<typeof personSchema>, Person>>
```
